store/postgres/asset_repository.go

```diff
@@ -0,0 +1,401 @@
+type AssetRepository struct {
+	client            *Client
+	userRepo          *UserRepository
+	defaultGetMaxSize int
+}
+
+// Get retrieves list of assets with filters via config
+func (r *AssetRepository) Get(ctx context.Context, config asset.Config) (assets []asset.Asset, err error) {
+	query, args := r.buildGetQuery(config)
+	ams := []*Asset{}
+	err = r.client.db.SelectContext(ctx, &ams, query, args...)
+	if err != nil {
+		err = fmt.Errorf("error getting asset list: %w", err)
+		return
+	}
+
+	assets = []asset.Asset{}
+	for _, am := range ams {
+		assets = append(assets, am.toAsset())
+	}
+
+	return
+}
+
+// Get retrieves list of assets with filters via config
+func (r *AssetRepository) GetCount(ctx context.Context, config asset.Config) (total int, err error) {
+	query, args := r.buildGetCountQuery(config)
+	err = r.client.db.GetContext(ctx, &total, query, args...)
+	if err != nil {
+		err = fmt.Errorf("error getting asset list: %w", err)
+	}
+
+	return
+}
+
+// GetByID retrieves asset by its ID
+func (r *AssetRepository) GetByID(ctx context.Context, id string) (ast asset.Asset, err error) {
+	query := `SELECT * FROM assets WHERE id = $1 LIMIT 1;`
+
+	am := &Asset{}
+	err = r.client.db.GetContext(ctx, am, query, id)
+	if errors.Is(err, sql.ErrNoRows) {
+		err = asset.NotFoundError{AssetID: id}
+		return
+	}
+	if err != nil {
+		err = fmt.Errorf("error getting asset with ID = \"%s\": %w", id, err)
+		return
+	}
+	ast = am.toAsset()
+
+	owners, err := r.getOwners(ctx, id)
+	if err != nil {
+		err = fmt.Errorf("error getting asset's owners with ID = \"%s\": %w", id, err)
+		return
+	}
+	ast.Owners = owners
+
+	return
+}
+
+// Upsert creates a new asset if it does not exist yet.
+// It updates if asset does exist.
+// Checking existance is done using "urn", "type", and "service" fields.
+func (r *AssetRepository) Upsert(ctx context.Context, ast *asset.Asset) error {
+	assetID, err := r.getID(ctx, ast)
+	if err != nil {
+		return fmt.Errorf("error getting asset ID: %w", err)
+	}
+	if assetID == "" {
+		assetID, err = r.insert(ctx, ast)
+		if err != nil {
+			return fmt.Errorf("error inserting asset to DB: %w", err)
+		}
+	} else {
+		err = r.update(ctx, assetID, ast)
+		if err != nil {
+			return fmt.Errorf("error updating asset to DB: %w", err)
+		}
+	}
+
+	ast.ID = assetID
+	return nil
+}
+
+// Delete removes asset using its ID
+func (r *AssetRepository) Delete(ctx context.Context, id string) error {
+	query := `DELETE FROM assets WHERE id = $1;`
+	res, err := r.client.db.ExecContext(ctx, query, id)
+	if err != nil {
+		return fmt.Errorf("error deleting asset with ID = \"%s\": %w", id, err)
+	}
+	affectedRows, err := res.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("error getting affected rows: %w", err)
+	}
+	if affectedRows == 0 {
+		return asset.NotFoundError{AssetID: id}
+	}
+
+	return nil
+}
+
+func (r *AssetRepository) buildGetQuery(config asset.Config) (query string, args []interface{}) {
```

Comment:
maybe using `https://github.com/Masterminds/squirrel` would make it more readable?

I was trying using this before with tag template story but the query doesn't really that complex like this one stew.
@StewartJingga

Reply:
nice library 👌